sms bolus from AAPSClient

mushroom-dev opened the pull request:
Implement bolus sms feature in AAPS client, if feature is enabled in menu and target phone number is provided it allows to send SMS directly from insulin/calc wizard dialog instead of "record only" feature. It still requires authorization code to be typed manually in the second message. If calc wizard was used, results will be saved in "treatment" tab history.

mushroom-dev committed:
sms command to send insulin wip

mushroom-dev committed:
- fix compilation issues
- preferences config update

mushroom-dev committed:
- working dialog from insulindialog.kt
- TODO: calc wizard doesn't send sms, remove animation

mushroom-dev committed:
- calc wizard working
- create calc entry in history if sms was send

mushroom-dev committed:
cleanup and formatting

mushroom-dev committed:
- improve sms option placement in menu
- fix this option to work only on aapsclient, only when it's enabledd

mushroom-dev commented:
Implementes #3507 and #3433

mushroom-dev committed:
fix some sonarscan issues

olorinmaia commented on app/src/main/res/values/strings.xml:
```diff
@@ -47,6 +47,9 @@
     <string name="a11y_dialog">dialog</string>
+    <!-- AAPS client sms -->
+    <string name="sms_receiver_number_dialog">If number is provided, insulin and calculator buttons will be able to send SMS command to inject bolus. Note that you still need to confirm it and reply with security token/code. Note that you need to enabled SMS control of AAPS phone to use this function.</string>
+    <string name="client_allow_sms">Allow AAPS client to send SMS with bolus command</string>
```
Typo here, AAPSClient instead of 'AAPS client'

olorinmaia commented on core/ui/src/main/res/values/strings.xml:
```diff
@@ -446,9 +446,11 @@
     <string name="cobvsiob">COB vs IOB</string>
     <string name="slowabsorptiondetected"><![CDATA[<font color=\'%1$s\'>!!!!! Slow carbs absorption detected: %2$d%% of time. Double check your calculation. COB can be overestimated thus more insulin could be given !!!!!</font>]]></string>
     <string name="partialboluswizard">Deliver this part of bolus wizard result [%]</string>
+    <string name="sms_receiver_number">Receiver phone number for sms commands send from AAPS client phone</string>
```
Some typos here.
Sent instead of send
AAPSClient instead of AAPS client

mushroom-dev committed:
review fixes

olorinmaia commented:
One concern is that AAPSClient doesn't always sync perfectly with AAPS. Both IOB and COB can be wrong due to many different reasons. We should consider adding a warning about this in the confirmation box of the bolus calculator in AAPSClient if enabling this feature since bolusing based on IOB/COB from AAPSClient's bolus calculator will then be more tempting to use when it's more accessible. 

I don't think there is a way with SMS commands to verify what IOB and COB AAPS really got. It's rare for my part that AAPSClient and AAPS got different IOB/COB, because if sync settings and NS is setup correctly, it's very unlikely to happen. But when looking at threads for support in FB/Discord/Github regarding wrong sync between AAPS and AAPSClient this is a concern I have. 

@MilosKozak In the OpenAPS log you can click to view in AAPSClient you can see the exact COB from AAPS; regardless if COB in NS / AAPSClient is wrong. Would it be possible to add IOB amount to the openAPS-log also? That way there is one place one could verify that COB/IOB in AAPSClient is correctly synced with AAPS without visiting master phone.

mushroom-dev commented on core/objects/src/main/kotlin/app/aaps/core/objects/wizard/BolusWizard.kt:
```diff
@@ -536,13 +545,18 @@ class BolusWizard @Inject constructor(
-                                if (!result.success) {
-                                    uiInteraction.runAlarm(result.comment, rh.gs(app.aaps.core.ui.R.string.treatmentdeliveryerror), app.aaps.core.ui.R.raw.boluserror)
+                        var phoneNumber = preferences.get(StringKey.SmsReceiverNumber)
+                        if (preferences.get(BooleanKey.SmsAllowRemoteCommands) && !phoneNumber.isNullOrBlank()) {
```
I will try to encapsulate it somehow

mushroom-dev replied:
I give up, I don't have time to resolve all the issues with gradle and dependencies during build

mushroom-dev committed:
add warning regarding IOB synchronization

mushroom-dev commented:
@olorinmaia  @MilosKozak what are next steps?

olorinmaia commented:
I'm not able to test this as my test-rig don't have sim-cards. One way is to inform about PR on Discord and see if there is any testers there and get more input regarding this PR generally.

mushroom-dev commented:
@MilosKozak can your review it? I assume groups of people building dev branches and parents needing features I implemented are disjoint. I asked about feedback on parents channel, didn't get any feedback. I have been testing it myself for some time.

emilisev commented:
I am testing this on top and beta1 and first use went all right. I will continue to test in the next few days.

emilisev commented:
When using the bolus assistant or quickwizard, the carbs are lost !
![Screenshot_20241208-124246](https://github.com/user-attachments/assets/903f7101-d80a-407d-888e-bff0568d017b)
![Screenshot_20241208-124437](https://github.com/user-attachments/assets/1b69d7da-ec7d-4524-a6ac-a564b35139f7)

mushroom-dev committed:
- move sms related settings to Button menu
- enable phone number verification in settings
- fix issues
- update AAPS response for SMS bolus request to include IOB

mushroom-dev commented:
@emilisev thanks for catching that! See my latest update for fixes, I found few other corner cases. I also moved sms related settings under button menu as you suggested.

@olorinmaia  I added current IOB to sms request response.

I've tested AAPSClient, haven't testes AAPS with updated SMS response yet, but I don't expect any issues there

emilisev commented:
I can't build it, I get : e: file:AndroidAPS-dev/plugins/main/src/main/kotlin/app/aaps/plugins/main/general/smsCommunicator/SmsCommunicatorPlugin.kt:936:65 Unresolved reference 'plus' for operator '+'.

mushroom-dev committed:
compilation fix

mushroom-dev commented:
I've pushed compilation fix. Note I didn't test it functionally yet. You are free to do it or you can wait for me. I will send update

mushroom-dev commented:
@emilisev I did some testing, haven't found any issues

emilisev commented:
Builds all right and carbs are not lost anymore, but I have encountered another case which seems problematic. If I use the bolus wizard from aapsclient but the bolus fails (previous bolus too close or wrong pin....), the carbs are still entered. I think we should be careful with that, it would be an issue is the parent misses the carbs entry, enters again the bolus wizard and the carbs are entered twice... I think this is quite risky...

mushroom-dev commented:
Personally I am fine with it. It works the same way with bolus alerts. Even if bolus fail timer is not stopped, if you provide another bolus with alarm then you have 2 countdowns. Or maybe it could be improved too? I don't think there's easy way to record carbs only after insulin was successfully administered. 

Maybe I could create some job which awakes every 20s to check if there's new carb entry in treatments tab, but I am not sure if such mechanism fits AAPS workflow. Also in that case some other person could enter carbs on primary phone, such scenario is also technically possible.

Personally as caregiver I am fine with current solution, if bolus fails I can always edit treatments tab, it's still big improvement for me. I will need to find time to fix that countdown issue though :)

gerison77 commented:
I tested it on 3.3.0-rc1 and it works as expected. I wouldn't miss that feature. It is a big improvement. Especially when you have to kids to remote bolus sometimes. Furthermore it helps to check the calculations afterwards.

I want to express my urgent wish this to become a feature in the standard release!

mushroom-dev commented:
I will try to postpone carbs entry after receiving confirmation about bolus delivery

emilisev commented:
I've been using the functionality almost on a daily basis for the last 2 weeks and it's growing on me :-) I never used the sms commands before, and just started using it to help test your functionality, but I have to admit it is quite useful!
For the "carb when bolus failed" behavior, I guess you get used to how it works. You should probably get more feedback on the matter, not just my own opinion :-)

mushroom-dev committed:
Merge remote-tracking branch 'upstream/dev' into mushroom/sms_bolus

mushroom-dev committed:
WIP extend sms bolus command

mushroom-dev committed:
Merge remote-tracking branch 'upstream/dev' into mushroom/sms_bolus

mushroom-dev committed:
- Introduce new BOLUSCARBS command
- Implement alarm support for CARBS sms command
- Implement regexp based validation for sms commands
- Tests for new features

mushroom-dev committed:
Merge remote-tracking branch 'upstream/dev' into mushroom/sms_bolus

mushroom-dev committed:
don't allow to set alarm on aapsclient's phone, minor fixes

mushroom-dev commented:
Sorry for long delay. I've made quite a lot of changes. I introduced new sms command "boluscarbs", as it's name indicates you can now send single sms with information about insulin and carbs. It's processed together so there should be no issue with carbs entry when insulin delivery fails. It allows also to set remote alarm on AAPS phone. I've made also few other minor changes

mushroom-dev committed:
- fix double response for boluscarbs command
- add missing file

mushroom-dev committed:
Merge remote-tracking branch 'upstream/dev' into mushroom/sms_bolus

mushroom-dev committed:
Handle bolus advisor (exit with explanation)

mushroom-dev committed:
Merge remote-tracking branch 'upstream/dev' into mushroom/sms_bolus

olorinmaia commented on core/ui/src/main/res/values/strings.xml:
```diff
@@ -454,13 +454,16 @@
     <string name="partialboluswizard_reset_time">Old glycemia time threshold [min]</string>
     <string name="bolus_constraint_applied_warn">Bolus constraint applied: %1$.2f U to %2$.2f U</string>
     <string name="bolus_recorded_only">Bolus will be recorded only (not delivered by pump)</string>
+    <string name="sms_request_notification">SMS command will be send</string>
```
"SMS command will be send" seem to be bad grammar.

"SMS command will be sent" or "The SMS command will be sent" is probably better if I understand the context. :)

mushroom-dev committed:
Support for comma separator in boluscarbs command
Fix typo

mushroom-dev committed:
Merge remote-tracking branch 'upstream/dev' into mushroom/sms_bolus

gerison77 commented:
I have been testing it now for one week on two rigs. Since the most recent update, I experienced no problems. I it a great improvement. Saves time and accurate entries of carbs.

mushroom-dev committed:
Merge remote-tracking branch 'upstream/dev' into mushroom/sms_bolus

mushroom-dev committed:
tests impl

mushroom-dev commented:
@MilosKozak it's ready for your review.

MilosKozak commented on core/keys/src/main/kotlin/app/aaps/core/keys/BooleanKey.kt:
```diff
@@ -70,7 +70,7 @@ enum class BooleanKey(
     AutotuneAdditionalLog("autotune_additional_log", false),
 
-    SmsAllowRemoteCommands("smscommunicator_remotecommandsallowed", false),
+    SmsAllowRemoteCommands("smscommunicator_remotecommandsallowed", false, showInApsMode = false, showInPumpControlMode = false),
```
Why this?

mushroom-dev replied:
This plugin is intended only for AAPSClient. It sends SMS commands to phone running AAPS. Both AAPS and control app are on the pump controlling phone so presence of that plugin would be confusing.

MilosKozak commented on core/objects/src/main/kotlin/app/aaps/core/objects/wizard/BolusWizard.kt:
```diff
@@ -86,9 +91,12 @@ class BolusWizard @Inject constructor(
     @Inject lateinit var processedDeviceStatusData: ProcessedDeviceStatusData
 
     var timeStamp: Long
+    // var phoneNumber: String
```
review https://github.com/nightscout/AndroidAPS/pull/3507/files and cleanup the code

MilosKozak commented on core/data/src/main/kotlin/app/aaps/core/data/configuration/Constants.kt:
```diff
@@ -11,7 +11,7 @@ object Constants {
 
     // SMS COMMUNICATOR
-    const val remoteBolusMinDistance = 15 * 60 * 1000L
+    const val remoteBolusMinDistance = 5 * 60 * 1000L
```
Why?

mushroom-dev replied:
15 minutes seems to be too much. In my case I've set some max carbs limit for single meal. It's intended to limit bolus to make sure carbs absorption kicks in on time. When meal is bigger than max carb then I split calculations into 2 meal with 5 minutes delay. 5 should also prevent accidental bolus at the same time from multiple caregivers

gerison77 replied:
Totally agree. I have the same process in place with the treatment on my two kids. Set bolus limit for safety, leading to mandatory split bolus for regular meals.

mushroom-dev committed:
cleanup

mushroom-dev committed:
Merge remote-tracking branch 'upstream/dev' into mushroom/sms_bolus

mushroom-dev committed:
Merge remote-tracking branch 'upstream/dev' into mushroom/sms_bolus

mushroom-dev committed:
Merge remote-tracking branch 'upstream/dev' into mushroom/sms_bolus

mushroom-dev commented:
@MilosKozak can you review again?

mushroom-dev committed:
Merge remote-tracking branch 'upstream/dev' into mushroom/sms_bolus

mushroom-dev committed:
Merge remote-tracking branch 'upstream/dev' into mushroom/sms_bolus

gerison77 commented:
I have tested it now on 3.3.3.0-dev since it's release without any issues on two rigs.  

From a practical point of view, being able to track the bolus calculation via bolus calculator in the treatments tab proofs valuable in order to draw conslusions of futher treatments and for general analysis. Correct SMS commands proof to save time, especially since it prevents typos which occur quite often. 

I furher suggest to extend the funcionalty on pump commands, Connect/Disconnect Pump, Stop/Enable loop. Which I have to use regualrily on sport aktivities like supervising my kids on Karate Trainings, MTB Tours where the setting of temp targets or % of Porfile just doesn't suffice to handle the situation. I am looking forward to have this functionality as well.

mushroom-dev committed:
Merge remote-tracking branch 'upstream/dev' into mushroom/sms_bolus

mushroom-dev committed:
Merge remote-tracking branch 'upstream/dev' into mushroom/sms_bolus

mushroom-dev committed:
merge fix

mushroom-dev committed:
Merge remote-tracking branch 'upstream/dev' into mushroom/sms_bolus
